File: examples/python/full_sign_control.py
"""Portable example. Replace InMemorySignService with the live provider in an Endstone plugin."""
from endstone_sign import (
    CardinalDirection,
    InMemorySignService,
    SignCloneRequest,
    SignKind,
    SignLocation,
    SignMaterial,
    SignMoveRequest,
    SignPatch,
    SignPlaceRequest,
    SignText,
    SignTextPatch,
    make_wall_hanging_sign_states,
    sign_block_identifier,
)

service = InMemorySignService()
shop = SignLocation("overworld", 100, 70, 100)
backup = SignLocation("overworld", 101, 70, 100)
relocated = SignLocation("overworld", 102, 70, 100)

placed = service.place(SignPlaceRequest(
    location=shop,
    block_identifier=sign_block_identifier(SignMaterial.CHERRY, SignKind.WALL_HANGING),
    states=make_wall_hanging_sign_states(CardinalDirection.NORTH),
    front=SignText(lines=("The Kingdom", "Market", "", ""), glowing=True),
    back=SignText(lines=("Staff Only", "", "", "")),
))
assert placed.ok

snapshot = service.capture(shop)
assert snapshot is not None
changed = service.apply(SignPatch(
    location=shop,
    expected_revision=snapshot.revision,
    front=SignTextPatch(
        line_updates={2: "Open 24/7"},
        argb=0xFFFFAA00,
        hide_glow_outline=True,
    ),
    waxed=True,
))
assert changed.ok

assert service.clone_sign(SignCloneRequest(shop, backup)).ok
assert service.move_sign(SignMoveRequest(backup, relocated)).ok

File: examples/cpp/full_sign_control.cpp
#include "endstone_sign/in_memory_adapter.h"
#include "endstone_sign/placement.h"
#include "endstone_sign/service.h"

#include <cassert>
#include <memory>

using namespace endstone_sign;

int main() {
    auto adapter = std::make_shared<InMemorySignAdapter>();
    SignService service(adapter);

    const SignLocation location{"overworld", 100, 70, 100};
    SignPlaceRequest place;
    place.location = location;
    place.block_identifier = signBlockIdentifier(SignMaterial::Cherry, SignKind::WallHanging);
    place.states = makeWallHangingSignStates(CardinalDirection::North);
    place.front.lines = {"The Kingdom", "Market", "", ""};
    place.back.lines = {"Staff Only", "", "", ""};
    assert(service.place(place).ok());

    const auto snapshot = service.capture(location);
    assert(snapshot);

    SignPatch patch;
    patch.location = location;
    patch.expected_revision = snapshot->revision;
    patch.front.emplace();
    patch.front->line_updates[2] = "Open 24/7";
    patch.front->argb = 0xFFFFAA00u;
    patch.front->glowing = true;
    patch.waxed = true;
    assert(service.apply(patch).ok());
}
